### PR TITLE
BB2-2200: Terraform BB2 BFD-Insights components part2

### PR DESCRIPTION
**Pull request by dtisza1**
**JIRA Ticket:**
[BB2-2200](https://jira.cms.gov/browse/BB2-2200)

**User Story or Bug Summary:**

Terraforming the **current** configuration of BB2 components for BFD Insights. This is in order to better manage resources and to preserve configuration in the event of a "disaster".

This includes Terraform for the **part 2 (of 3)** related Jira stories. This is split out from the larger DRAFT PR [here](https://github.com/CMSgov/beneficiary-fhir-data/pull/1777).

#### NOTES:

-  Developing using the `allenv` workspace since this is a shared/common component. We will work to split this out in to `test,impl,prod` environments at a later date and not as part of this work.
- The resources are imported in to the state for the `allenv` workspace currently.

- The Terraform plan is showing the expected changes and they will be applied after PR review is complete!

#### The following is a summary of work done for this PR:

- The `default_tags` under the `bb2/services/common` path are:
  - business = "OEDA"
  - application = "bfd-insights"
  - project = "bb2"
  - Environment = "allenv"
  - stack = "bfd-insights-bb2-allenv"

- Terraformed the following common related component under `bb2/services/common`:
  - Uses Terraform workspace: `allenv`
  - Default state key:  bfd-insights/bb2/services/common/terraform.tfstate
  - Resources managed under `modules/lambda`:
    - resource "aws_lambda_function" "firehose_log_processor_lambda"
    - This lambda is used for processing events sent to the Firehose by Cloudwatch Logs subscription filters.
  - A slightly modified Amazon provided Python lambda function is located at: `bb2/services/common/modules/lambda/lambda_src/lambda_function.py`
    - This processes events sent to the Kinesis Firehose by Cloudwatch log subscription filters.
    - It is modified to flatten JSON related to the BB2 log structure by pulling fields up a level that are under the `message.*` path.

- Added `required_providers` to specify versions to be used for `common/` and `log_stream/` services.
  - Required versions are:  `hashicorp/aws` = `~> 5.7.0` and `hashicorp/archive` = `~> 2.4.0`.

- Updated the top level `README.md` with overview and usage instruction.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: insights/terraform/projects/bb2/services/common/modules/lambda/main.tf [hunk 1]
```diff
@@ -0,0 +1,52 @@
+data "aws_caller_identity" "current" {}
+
+data "aws_kms_key" "bucket_cmk" {
+  key_id = "alias/bfd-insights-bb2-cmk"
+}
+
+# Zip File containing Lambda script
+data "archive_file" "zip-archive-format-firehose-logs" {
```
Comment (billie-rose) [inline insights/terraform/projects/bb2/services/common/modules/lambda/main.tf:8]:
```suggestion
data "archive_file" "zip_archive_format_firehose_logs" {
```

File: insights/terraform/projects/bb2/services/common/modules/lambda/variables.tf [hunk 2]
```diff
@@ -0,0 +1,15 @@
+variable "name" {
```
Comment (billie-rose) [inline insights/terraform/projects/bb2/services/common/modules/lambda/variables.tf:1]: Can we add descriptions to these variables?

File: insights/terraform/projects/bb2/services/common/modules/lambda/main.tf [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  memory_size                    = 128
+  package_type                   = "Zip"
+  reserved_concurrent_executions = -1
+  role                           = local.lambda_role_arn
+  publish                        = false
+  runtime                        = "python3.8"
+  source_code_hash               = data.archive_file.zip-archive-format-firehose-logs.output_base64sha256
+
+  tags = {
+    "lambda-console:blueprint" = "kinesis-firehose-cloudwatch-logs-processor-python",
```
Comment (billie-rose) [inline insights/terraform/projects/bb2/services/common/modules/lambda/main.tf:42]:
```suggestion
    "lambda-console:blueprint" = "kinesis-firehose-cloudwatch-logs-processor-python"
```

File: insights/terraform/projects/bb2/services/common/main.tf [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    project     = local.project
+    Environment = local.env
+    stack       = "${local.application}-${local.project}-${local.env}"
+  }
+  application = "bfd-insights"
+  project     = "bb2"
+  env         = terraform.workspace
+  region      = "us-east-1"
+
+  # Shared lambda name/role (TODO: Split out by env)
```
Comment (billie-rose) [inline insights/terraform/projects/bb2/services/common/main.tf:14]: Is there a story or ticket we can reference here for this TODO?

File: insights/terraform/projects/bb2/services/common/backend.tf [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      source  = "hashicorp/aws"
+      version = "~> 5.7.0"
+    }
+
+    archive = {
+      source  = "hashicorp/archive"
+      version = "~> 2.4.0"
+    }
+  }
+}
```
Comment (billie-rose) [inline insights/terraform/projects/bb2/services/common/backend.tf:29]:
```suggestion
}

```

File: insights/terraform/projects/bb2/README.md [hunk 6]
```diff
@@ -6,11 +6,16 @@ The BB2 BFD-Insights AWS components are setup under this directory.
...
 
+The Terraform workspaces feature is used for switching between `test/impl/prod/allenv` environments. The `allenv` workspace is for components that are shared between `test/impl/prod` environments.
+
 The BB2 specific components are setup under the following service areas:
 
 - `services/analytics`:
   - Includes components used for reporting via QuickSight.
 
+- `services/common`:
+  - Includes components that are currently common to `test/impl/prod` enviornments. Uses the `allenv` workspace.
```
Comment (billie-rose) [inline insights/terraform/projects/bb2/README.md:17]:
```suggestion
  - Includes components that are currently common to `test/impl/prod` environments. Uses the `allenv` workspace.
```

File: insights/terraform/projects/bb2/services/common/modules/lambda/main.tf [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+data "aws_kms_key" "bucket_cmk" {
+  key_id = "alias/bfd-insights-bb2-cmk"
+}
+
+# Zip File containing Lambda script
+data "archive_file" "zip-archive-format-firehose-logs" {
+  type        = "zip"
+  source_file = "${path.module}/lambda_src/lambda_function.py"
+  output_path = "${path.module}/lambda_function.zip"
```
Comment (billie-rose) [inline insights/terraform/projects/bb2/services/common/modules/lambda/main.tf:11]: Should we add a `.gitignore` for zips here?

**Comment by billie-rose** [comment]
Just a few comments. Thank you!

File: insights/terraform/projects/bb2/services/common/main.tf (hunk 4 above)
Comment (dtisza1) [reply insights/terraform/projects/bb2/services/common/main.tf:14]: @billie-rose I created a story in our backlog for this to be groomed. TY!

File: insights/terraform/projects/bb2/services/common/modules/lambda/main.tf (hunk 7 above)
Comment (dtisza1) [reply insights/terraform/projects/bb2/services/common/modules/lambda/main.tf:11]: TY for catching this. Forgot to add the file before.

**Comment by dtisza1** [comment]
> Just a few comments. Thank you!

@billie-rose I just pushed up changes that I believe cover your change suggestions. Let me know if this looks OK and/or if you see more improvements. Thank you for the review on this!

**Review by oragame — APPROVED** [review]
Nice work, Dave! tested locally and expected plan is showing for the 'allenv' workspace.

**Review by mjburling — APPROVED** [review]
Thanks! This iterative approach is working for me, and I think this looks sound.

**Pull request merged**